x-pack/plugins/uptime/public/components/monitor/synthetics/waterfall/components/waterfall_chart.tsx

```diff
@@ -0,0 +1,221 @@
+  TooltipType,
+} from '@elastic/charts';
+import { EUI_CHARTS_THEME_DARK, EUI_CHARTS_THEME_LIGHT } from '@elastic/eui/dist/eui_charts_theme';
+// NOTE: The WaterfallChart has a hard requirement that consumers / solutions are making use of KibanaReactContext, and useKibana etc
+// can therefore be accessed.
+import { useUiSetting$ } from '../../../../../../../../../src/plugins/kibana_react/public';
+// NOTE: The WaterfallChart expects the ThemeProvider to have been setup for styled-components, using the EUI theme. However,
+// as some plugins aren't utilising the ThemeProvider, the WaterfallChart sets up the provider itself just incase.
+import { EuiThemeProvider } from '../../../../../../../observability/public';
+import { useWaterfallContext } from '../context/waterfall_chart';
+import {
+  WaterfallChartOuterContainer,
+  WaterfallChartFixedTopContainer,
+  WaterfallChartFixedTopContainerSidebarCover,
+  WaterfallChartFixedAxisContainer,
+  WaterfallChartSidebarContainer,
+  WaterfallChartSidebarContainerInnerPanel,
+  WaterfallChartSidebarContainerFlexGroup,
+  WaterfallChartSidebarFlexItem,
+  WaterfallChartChartContainer,
+  WaterfallChartLegendContainer,
+  WaterfallChartTooltip,
+} from './styles';
+import { WaterfallData } from '../types';
+
+const Tooltip = ({ header }: TooltipInfo) => {
+  const { data, renderTooltipItem } = useWaterfallContext();
+  const relevantItems = data.filter((item) => {
+    return item.x === header?.value;
+  });
+  return (
+    <WaterfallChartTooltip>
+      <EuiFlexGroup direction="column" gutterSize="none">
+        {relevantItems.map((item, index) => {
+          return (
+            <EuiFlexItem key={index}>{renderTooltipItem(item.config.tooltipProps)}</EuiFlexItem>
+          );
+        })}
+      </EuiFlexGroup>
+    </WaterfallChartTooltip>
+  );
+};
+
+export type RenderItem<I = any> = (item: I, index: number) => JSX.Element;
+
+interface WaterfallChartProps {
+  tickFormat: TickFormatter;
+  domain: DomainRange;
+  barStyleAccessor: BarStyleAccessor;
+  renderSidebarItem?: RenderItem;
+  renderLegendItem?: RenderItem;
+  maxHeight?: number;
+}
+
+const getUniqueBars = (data: WaterfallData) => {
+  return data.reduce<number[]>((acc, item) => {
+    if (acc.indexOf(item.x) === -1) {
+      acc.push(item.x);
+      return acc;
+    } else {
+      return acc;
+    }
+  }, []);
+};
+
+const BAR_HEIGHT = 32;
+const getChartHeight = (data: WaterfallData): number => getUniqueBars(data).length * BAR_HEIGHT;
+
+const MAIN_GROW_SIZE = 8;
+const SIDEBAR_GROW_SIZE = 2;
+
+export const WaterfallChart = ({
+  tickFormat,
+  domain,
+  barStyleAccessor,
+  renderSidebarItem,
+  renderLegendItem,
+  maxHeight = 600,
+}: WaterfallChartProps) => {
+  const { data, sidebarItems, legendItems } = useWaterfallContext();
+  const [darkMode] = useUiSetting$<boolean>('theme:darkMode');
+
+  const theme = useMemo(() => {
+    return darkMode ? EUI_CHARTS_THEME_DARK.theme : EUI_CHARTS_THEME_LIGHT.theme;
+  }, [darkMode]);
```

Comment:
i wish there was a better way to pass this to charts, it's done in every uptime chart.

Reply:
Yeah, a one-use context of some sort would be nice.